File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/validation/comparator/RegexValueComparator.java
package org.skellig.teststep.processing.validation.comparator;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

class RegexValueComparator implements ValueComparator {

    private static final Pattern PATTERN = Pattern.compile("regex\\((.+)\\)");
    private static final String REGEX_PREFIX = "regex(";

    @Override
    public boolean compare(Object expectedValue, Object actualValue) {
        if (actualValue != null) {
            Matcher matcher = PATTERN.matcher(String.valueOf(expectedValue));
            if (matcher.find()) {
                String regex = matcher.group(1);
                String actualValueAsString = String.valueOf(actualValue);
                return isMatchRegex(regex, actualValueAsString);
            }
        }
        return false;
    }

    private boolean isMatchRegex(String regex, String actualValueAsString) {
        try {
            Pattern expectedPattern = Pattern.compile(regex);
            return expectedPattern.matcher(actualValueAsString).matches();
        } catch (Exception ex) {
            return false;
        }
    }

    @Override
    public boolean isApplicable(Object expectedValue) {
        return String.valueOf(expectedValue).contains(REGEX_PREFIX);
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/validation/comparator/RegexValueComparator.kt
package org.skellig.teststep.processing.validation.comparator

import java.util.regex.Pattern

class RegexValueComparator : ValueComparator {

    companion object {
        private val PATTERN = Pattern.compile("regex\\((.+)\\)")
        private const val REGEX_PREFIX = "regex("
    }

    override fun compare(expectedValue: Any?, actualValue: Any?): Boolean {
        return actualValue?.let {
            val matcher = PATTERN.matcher(expectedValue.toString())
            if (matcher.find()) {
                val regex = matcher.group(1)
                val actualValueAsString = actualValue.toString()
                isMatchRegex(regex, actualValueAsString)
            } else {
                false
            }
        } ?: false
    }

    private fun isMatchRegex(regex: String, actualValueAsString: String): Boolean {
        return try {
            val expectedPattern = Pattern.compile(regex)
            expectedPattern.matcher(actualValueAsString).matches()
        } catch (ex: Exception) {
            false
        }
    }

    override fun isApplicable(expectedValue: Any?): Boolean {
        return (expectedValue?.toString() ?: "").contains(REGEX_PREFIX)
    }
}